feat: Add property-based authorization to OC REST API and CamundaClient

ce-dmelnych opened the pull request:
## Description

This PR provides support for property-based authorizations to the Orchestration Cluster Authorization REST API and the CamundaClient.

In addition to ID-based authorizations (`resourceId`), the API and client now support authorizations that are defined against a resource property via a new `resourcePropertyName` field. This enables more flexible, attribute-style access control while keeping the REST API structure explicit and easy to understand.

### What this PR changes

#### 1. REST API: Authorization models and validation

- Adds support for `resourcePropertyName` to the OC Authorization REST API models:
  - Create/Update authorization
  - Get/Search authorization
- Updates the OpenAPI spec (`authorization.yaml`) so that:
  - `AuthorizationRequest` is now defined as a `oneOf`:
    - `AuthorizationIdBasedRequest`
    - `AuthorizationPropertyBasedRequest` (new)
  - The two variants are mutually exclusive:
    - ID-based authorization requires `resourceId`.
    - Property-based authorization requires `resourcePropertyName`.
- Enforces validation for Create/Update endpoints:
  - Exactly one of `resourceId` or `resourcePropertyName` must be provided.
  - It is invalid to:
    - Provide both fields, or
    - Provide neither field / leave both empty.

For ID-based auth (keep the same JSON structure as before):
```json
{
  "ownerId": "string",
  "ownerType": "USER",
  "resourceId": "string",
  "resourceType": "AUTHORIZATION",
  "permissionTypes": ["ACCESS"]
}
```

For property-based auth (new):

```json
{
  "ownerId": "string",
  "ownerType": "USER",
  "resourcePropertyName": "string", // new property
  "resourceType": "USER_TASK",
  "permissionTypes": ["CLAIM"]
}
```
- Extends the Get/Search Authorizations response to include `resourcePropertyName` and to `sort`/`filter` by `resourcePropertyName/s`.

> [!NOTE]
> The [Get](https://docs.camunda.io/docs/next/apis-tools/orchestration-cluster-api-rest/specifications/get-authorization/) and [Search](https://docs.camunda.io/docs/next/apis-tools/orchestration-cluster-api-rest/specifications/search-authorizations/) Authorizations endpoints will omit `resourceId` and `resourcePropertyName` from the JSON response when those values are not relevant for the given authorization. Specifically, for resource-id–based authorizations, `resourcePropertyName` is not included in the response, even though it may be stored as an empty string (`""`) in secondary storage; similarly, for property-based authorizations, `resourceId` is omitted.
>
> This behavior is implemented by setting the corresponding fields to `null` in [SearchQueryResponseMapper#toAuthorization](https://github.com/camunda/camunda/pull/41079/files#diff-7e539eb0200aa8a710f73c73bfacbf8953f170ed44625db7a194bdc51fad7bff) when they are empty, so that responses are not overloaded with properties that do not apply to the current authorization.
> <details><summary>Examples</summary>
>
> <p><code>ID/ANY</code> authorization will look like:</p>
> <pre><code>
> {
>   "authorizationKey": "2",
>   "ownerId": "foo",
>   "ownerType": "USER",
>   "resourceType": "RESOURCE",
>   "resourceId": "*",
>   "permissionTypes": ["CREATE"]
> }
> </code></pre>
>
> <p>and a <code>PROPERTY</code> authorization will look like:</p>
> <pre><code>
> {
>   "authorizationKey": "3",
>   "ownerId": "foo",
>   "ownerType": "USER",
>   "resourceType": "USER_TASK",
>   "resourcePropertyName": "assignee",
>   "permissionTypes": ["CLAIM"]
> }
> </code></pre></details>

#### 2. CamundaClient support

- Extends the CamundaClient so that:
  - `newCreateAuthorizationCommand` supports specifying `resourcePropertyName` as an alternative to `resourceId` (mutually exclusive).
  - `newUpdateAuthorizationCommand` supports `resourcePropertyName` for updating existing authorizations (mutually exclusive with `resourceId`).
  - `newAuthorizationGetRequest` and `newAuthorizationSearchRequest` include `resourcePropertyName` in their response types.
- allows `newAuthorizationSearchRequest` to sort and filter by `resourcePropertyName`.

### Why this is needed
Property-based authorization will be used for the `USER_TASK` resource type. This allows defining authorization rules that grant permissions based on specific task properties (such as `assignee`, `candidateUsers`, or `candidateGroups`), rather than only by resource ID or wildcard (`*`).

### Commit structure
The commits in this PR are structured logically by concern to make the review process easier and to clearly show how each part of the stack is adapted for property-based authorization.

## Related issues

closes #40079

ce-dmelnych committed:
fix: correct schema references for createAuthorization endpoint

It seems that the recent refactoring that moved endpoint definitions from `rest-api.yaml`
into individual component files introduced incorrect schema references in `authorizations.yaml`.
The POST `/authorizations` endpoint mistakenly used `AuthorizationCreateResult` as the request body
and `AuthorizationResult` as the response schema.

This caused incorrect API documentation generation for the rendered OC REST API spec.

Updated the endpoint to use AuthorizationRequest as the request schema and
AuthorizationCreateResult as the response schema, restoring correct generation for endpoint:
https://docs.camunda.io/docs/8.9/apis-tools/orchestration-cluster-api-rest/specifications/create-authorization/

ce-dmelnych committed:
feat: add resourcePropertyName to authorization create/update requests

Introduce property-based authorizations via `resourcePropertyName` alongside existing
id-based (`resourceId`) ones by redefining `AuthorizationRequest` as a polymorphic (`oneOf`) schema
composed of `AuthorizationIdBasedRequest` and `AuthorizationPropertyBasedRequest`.

This enforces mutual exclusivity structurally and allows a single create/update endpoint to serve
both authorization modes without payload ambiguity.

ce-dmelnych committed:
feat: add AuthorizationRequest deserializer for id/property-based auth

Introduce and register custom deserializer for `AuthorizationRequest`
to distinguish between `AuthorizationIdBasedRequest` and `AuthorizationPropertyBasedRequest`
based on presence of either `resourceId` or `resourcePropertyName`.

Enforces mutual exclusivity and validates that exactly one is provided (rejects none or both),
enabling a single create/update endpoint to support both id-based and property-based
authorization models without schema ambiguity.

ce-dmelnych committed:
feat: recognize AuthorizationIdBasedRequest in RequestMapper validator

Refactor `RequestMapper` to switch on `AuthorizationRequest` and handle
`AuthorizationIdBasedRequest` (preparing for property-based variant).

Extract shared field validation into private helper in `AuthorizationRequestValidator` to enable
upcoming `AuthorizationPropertyBasedRequest` support without duplicating checks.

ce-dmelnych committed:
feat: recognize AuthorizationPropertyBasedRequest in mapper, validator

Adds mapping & validation for `AuthorizationPropertyBasedRequest` and adjusts `AuthorizationServices`
to convert `CreateAuthorizationRequest` and `UpdateAuthorizationRequest` into `BrokerAuthorizationRequest`
using the `PROPERTY` resource matcher and populating `resourcePropertyName` for property-based authorizations.

Also extends `AuthorizationControllerTest` with parameterized cases
to cover invalid property-based authorization requests.

ce-dmelnych committed:
test: parametrize authorization create/update tests

Parametrized the authorization create/update tests to validate all scenarios:
- resourceId: ID
- resourceId: ANY ("*")
- resourcePropertyName: PROPERTY.

Simplified test setup by removing full construction of AuthorizationRecord as this value is
returned by the mocked `authorizationServices#createAuthorization` and `updateAuthorization` methods.

Replaced field-by-field assertions with a single equality assertion and moved
`expectedCreateRequest` into the parameter provider to improve clarity.

ce-dmelnych committed:
refactor: move authorization resourceMatcher mapping to RequestMapper

Move the resolution of `AuthorizationResourceMatcher` value from `AuthorizationService` to `RequestMapper`
by including it in `CreateAuthorizationRequest` and `UpdateAuthorizationRequest` records.
This change centralizes the mapping logic in a single location, making it easier to
maintain and understand the transformation from REST API models to service layer and broker requests.

Additionally, the matcher value can now be directly verified in controller tests
(`AuthorizationControllerTest#createAuthorizationShouldReturnTheKey` and `#updateAuthorizationShouldReturnNoContent`)
by asserting the captured value, which itself increases test coverage.

ce-dmelnych committed:
feat: add resourcePropertyName to AuthorizationResult

Add support for property-based authorization in the `AuthorizationResult` schema.
The result now includes `resourcePropertyName`.

This change affects `GET /authorizations/{authorizationKey}` and `POST /authorizations/search endpoints`.

The `SearchQueryResponseMapper.toAuthorization` method now maps the `resourcePropertyName` field
from `AuthorizationEntity` to `AuthorizationResult` without transformation.

ce-dmelnych committed:
feat: omit empty resourceId/resourcePropertyName in AuthorizationResult

Adjusted mapping to `AuthorizationResult` to return `null` for empty ("") or `null`
`resourcePropertyName` and `resourceId` values. This prevents empty fields from
appearing in the JSON response.

As a result, queries for existing and new ID/ANY authorizations would produce the same JSON
as before (no `resourcePropertyName`), while new PROPERTY authorizations will include
only `resourcePropertyName` and omit `resourceId`.
This allows clients to reliably distinguish between id-based and property-based authorizations.

Cotroller tests adjusted to assert absence/presence of these fields.

ce-dmelnych committed:
feat: enable sorting and filtering by resourcePropertyName (REST-API)

Add support for sorting and filtering authorizations by `resourcePropertyName` property.
Extended dedicated mappers to handle `resourcePropertyNames` filter criteria and support sorting by
`resourcePropertyName` field.

Updated `AuthorizationQueryControllerTest` with parameterized test cases validating new query capabilities
for both sort and filter operations with `resourcePropertyName`.

ce-dmelnych committed:
feat: add resourcePropertyName to authorization commands (Client)

Enable creation and update of property-based authorizations via CamundaClient.
Both `newCreateAuthorizationCommand` and `newUpdateAuthorizationCommand` now support
`resourcePropertyName` as an alternative to `resourceId` in the same builder step,
enforcing mutual exclusivity at the API level.

Users must choose either `resourceId` or `resourcePropertyName` when constructing the AuthorizationRequest.

Extended `CreateAuthorizationTest` and `UpdateAuthorizationTest` with test cases
validating property-based authorization.

ce-dmelnych committed:
test: refactor AuthorizationIntegrationTest to prepare for prop auth

Introduce `AuthorizationRequestParam` record to encapsulate authorization request parameters
and enable parameterized testing for both ID-based and property-based authorizations.

The record includes factory methods `idBased()` and `propertyBased()` to simplify test data creation
and a helper method `isIdBased()` to distinguish between authorization types.

This refactoring prepares the test suite for upcoming property-based authorization test cases.

ce-dmelnych committed:
feat: add resourcePropertyName to Authorization model, mapper (Client)

Extend CamundaClient `Authorization` model with `resourcePropertyName`
and update `SearchResponseMapper` to map it.

Adapt create/update authorization integration tests to verify property-based authorizations.

ce-dmelnych committed:
feat: add resourcePropertyName to filter and sort models (Client)

Extend `AuthorizationFilter` with `resourcePropertyNames()` methods and
`AuthorizationSort` with `resourcePropertyName()` sorting option.

Add integration tests to verify filtering and sorting
by resource property names for property-based authorizations.

tmetzke reviewed (commented):
👍 API code looks good to me 🚀

koevskinikola reviewed (approved):
Authorization changes LGTM @ce-dmelnych 👍

npepinpe reviewed (approved):
Just some questions, but no blocker. Please just address the part about the breaking change. I guess you already thought about it and so on, but I'm curious why it's not a breaking API change.

ce-dmelnych committed:
refactor: use Collection for authorization id and propertyName filters

The order of elements used to filter authorizations does not matter, so the authorization filters
now accepts `Collection<String>` (alongside varargs overloads) for `resourceIds` and `resourcePropertyNames`
to provide a more flexible API for callers.

Changing the existing `resourceIds(List<String>)` signature was reported by Revapi plugin
as a removed method, the change is effectively backward-compatible for users
since any `List` also implements `Collection`.
A corresponding Revapi ignore entry was added to document this adjustment.

ce-dmelnych committed:
refactor: inline nullIfEmpty helper in search response mapper

Inline the `nullIfEmpty` helper in `SearchQueryResponseMapper` since it was used only twice
and did not contain any complex logic.

ce-dmelnych commented:
Updated branch with rebase to get rid of strange compilation issues on CI.
